### DISCUSSION TITLE
SvgViewbox ignores the whitespace around the image

### DISCUSSION THREAD

**Discussion by jaytonic** [ElinamLLC/SharpVectors#191] [Q&A]
Hi,

I've 2 SVG that I want to overlay.

One is the main image, always displayed, taking the full height/width.
The second one(linked) is only on the bottom of the image. But when I superpose them, it gets displayed on the middle:

```xml
<Grid>
  <svgc:SvgViewbox Source="SingleExpansionCollar.svg"  />
  <svgc:SvgViewbox Stretch="None" Source="SingleExpansionCollarBottom.svg"
         Visibility="{Binding IsBottomPartVisible, Converter={StaticResource BooleanToVisibilityConverter}}"/>   
</Grid>
```
Is there a way to have the second image(...Bottom.svg) that takes the full space without trimming the whitespace?

[CollarSingle.zip](https://github.com/ElinamLLC/SharpVectors/files/6539843/CollarSingle.zip)

**Comment by paulushub** [comment]
@jgrossrieder Please, can you draw an picture/illustration of how you want the final result to look like? Just render each, copy the screenshots and position with any drawing application like MS-Paint.

**Reply by jaytonic** [reply]
![image](https://user-images.githubusercontent.com/7582002/119514792-9b582300-bd75-11eb-9c0f-738b849ce10a.png)
Here is how it should look like

**Comment by paulushub** [comment]
@jgrossrieder There are several layout combination to get that and the results depends on the uses.
Here is a sample using `StackPanel` and `Image` converters
```xml
<Window x:Class="WpfApp4.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:svgc="http://sharpvectors.codeplex.com/svgc/"
        xmlns:local="clr-namespace:WpfApp4"
        mc:Ignorable="d"
        Title="MainWindow" Height="720" Width="400">
    <StackPanel Orientation="Vertical" HorizontalAlignment="Center">
        <Image Source="{svgc:SvgImage Source=SingleExpansionCollar.svg, AppName=WpfApp4}"/>
        <Image Source="{svgc:SvgImage Source=SingleExpansionCollarBottom.svg, AppName=WpfApp4}" Margin="100 50 20 10"/>
    </StackPanel>
</Window>
```
Here I used the margin property on the bottom image to both constrain the position and size. The results is shown below:

![shots](https://user-images.githubusercontent.com/809351/119523622-ded38100-bdb7-11eb-9ea6-d1f2f1f02aca.png)

The same results with `SvgViewBox`, removing the `Stretch` property on the bottom image, and using the `Margin` instead.
```xml
<Window x:Class="WpfApp4.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:svgc="http://sharpvectors.codeplex.com/svgc/"
        xmlns:local="clr-namespace:WpfApp4"
        mc:Ignorable="d"
        Title="MainWindow" Height="720" Width="400">
    <StackPanel Orientation="Vertical" HorizontalAlignment="Center">
        <svgc:SvgViewbox Source="SingleExpansionCollar.svg"/>
        <svgc:SvgViewbox Source="SingleExpansionCollarBottom.svg" Margin="100 50 20 10"/>
    </StackPanel>
</Window>
```
Or using the `DockPanel` as shown below:
```xml
<Window x:Class="WpfApp4.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:svgc="http://sharpvectors.codeplex.com/svgc/"
        xmlns:local="clr-namespace:WpfApp4"
        mc:Ignorable="d"
        Title="MainWindow" Height="720" Width="400">
    <DockPanel HorizontalAlignment="Center">
        <svgc:SvgViewbox Source="SingleExpansionCollarBottom.svg" Margin="200 50 20 50" DockPanel.Dock="Bottom"/>
        <svgc:SvgViewbox Source="SingleExpansionCollar.svg"/>
    </DockPanel>
</Window>
```
And finally, a little different layout using the `Grid`:
```xml
<Window x:Class="WpfApp4.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:svgc="http://sharpvectors.codeplex.com/svgc/"
        xmlns:local="clr-namespace:WpfApp4"
        mc:Ignorable="d"
        Title="MainWindow" Height="620" Width="400">
    <Grid HorizontalAlignment="Center" VerticalAlignment="Center">
        <Grid.RowDefinitions>
            <RowDefinition/>
            <RowDefinition/>
        </Grid.RowDefinitions>
        <svgc:SvgViewbox Source="SingleExpansionCollar.svg" Grid.Row="0" Margin="10"/>
        <svgc:SvgViewbox Source="SingleExpansionCollarBottom.svg" Grid.Row="1" HorizontalAlignment="Right" 
                       Height="60" Margin="10 30"/>
    </Grid>
</Window>
```

**Reply by jaytonic** [reply]
Thank you for your answer. I've seen that I can play with margin to do the correct placement. 
But I've like 100 images, and I've the same thing for the top, bottom left and right, if I must play with each image to have the correct margin, it's a nightmare.


(I cannot remove the Stretch because the small "bottom" part becomes too big)
Also, Grid with row definitions, stack panels and dockpanel doesn't work, because when I will have the top element present, It will shift the other images on the bottom. I need some kind of absolute placement.


Isn't there a way to display the image like inkscape/browsers do? Taking into account the whitespace?

**Reply by paulushub** [reply]
Definition of the [Canvas control](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.canvas?view=net-5.0)
> Defines an area within which you can **explicitly position** child elements by using coordinates that are relative to the Canvas area.

Use `FileSvgReader` to convert the drawing to `DrawingGroup` and wrap each in `DrawingImage`,  then use the WPF `Canvas` to position and display.

Alternatively, you can have a main `DrawingGroup`, apply `TranslateTransform` to each converted `DrawingGroup`, add it to the main drawing group as a child and use `SvgDrawingCanvas` to display the result.

**Reply by jaytonic** [reply]
Yes, my current solution uses a canvas where I manually put the margin for everything. But, the image contains some whitespace around, when drawing this image at position x,y, IMHO, it should not trim the whitespace(or there should be an option to not trim it).

We have a lot of other logic that we bind to each individual `SvgViewbox`(like brushes, visibility, ...), and each element won't have the same value

**Reply by paulushub** [reply]
I guess by "whitespace" you mean some bounding box. For such, use the following properties of `WpfDrawingSettings` to see which works for your drawings
1. **IgnoreRootViewbox**, **false** by default
2. **EnsureViewboxSize**, **false** by default
3. **EnsureViewboxPosition**, **true** by default - might be the source of it.

**Reply by jaytonic** [reply]
Yeah, you're probably right about the bounding box, sorry. 

Regarding the `WpfDrawingSettings`, is there a way to configure it in the XAML? I'm searching, but I'm only finding example where used on the c# side. Some of its settings(like OptimizePath) are available as properties on the SvgViewbox, but I didn't found anything for `EnsureViewboxPosition`

**Reply by paulushub** [reply]
Sorry, these properties are not exposed by the controls, but by the converters.
Please try using one of the converters to test if that is the solution. I will expose those properties
in the controls in the next update.

**Reply by jaytonic** [reply]
I'm struggling on the dummest thing: I can't find why I get ArgumentException here when I pass the filename of my SVG to the converter. It apparently doesn't find it.

**Reply by jaytonic** [reply]
I've managed to load the image directly from a disk path for testing after seeing it was not possible to provide a file in the resources.

I've tried the following:

```
            WpfDrawingSettings settings = new WpfDrawingSettings();
            settings.EnsureViewboxPosition = false;
            settings.IncludeRuntime = true;
            StreamSvgConverter converter = new StreamSvgConverter(settings);
            MemoryStream memoryStream = new MemoryStream();

            if (converter.Convert(@"F:\SingleExpansionCollarBottom.svg", memoryStream))
            {
	            BitmapImage bitmap = new BitmapImage();
	            bitmap.BeginInit();
	            bitmap.CacheOption = BitmapCacheOption.OnLoad;
	            bitmap.StreamSource = memoryStream;
	            bitmap.EndInit();
	            bottom.Source = bitmap;
            }
            memoryStream = new MemoryStream();
            if (converter.Convert(@"F:\SingleExpansionCollarTop.svg", memoryStream))
            {
	            BitmapImage bitmap = new BitmapImage();
	            bitmap.BeginInit();
	            bitmap.CacheOption = BitmapCacheOption.OnLoad;
	            bitmap.StreamSource = memoryStream;
	            bitmap.EndInit();
	            top.Source = bitmap;
            }
```

in xaml: 
```

		<Image Stretch="None" Name="bottom"  />
		<Image Stretch="None" Name="top"    />
```

But nothing gets displayed now :/(no error either)

**Comment by paulushub** [comment]
@jaytonic I have just tried looking at the files and there is no extra spaces added. For the sample below, I added border to see the extend of the images: Maybe `SvgViewbox`, which is an extension of the WPF `Viewbox` is not the right choice of control for your needs.

```xml
<Window x:Class="WpfApp4.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:svgc="http://sharpvectors.codeplex.com/svgc/"
        xmlns:local="clr-namespace:WpfApp4"
        mc:Ignorable="d"
        Title="MainWindow" Height="420" Width="600">
    <Canvas>
        <Border BorderBrush="Red" BorderThickness="1" Canvas.Left="10" Canvas.Top="10">
            <Image x:Name="CallarBottom" Stretch="None" 
                    Source="{svgc:SvgImage Source=SingleExpansionCollar.svg, AppName=WpfApp4}"/>
        </Border>            
        <Border BorderBrush="Red" BorderThickness="1" Canvas.Left="100" Canvas.Top="100">
            <Image Stretch="None" Source="{svgc:SvgImage Source=SingleExpansionCollarBottom.svg, AppName=WpfApp4}"/>
        </Border>            
    </Canvas>
</Window>
```

You can convert to image using `new DrawingImage(drawingGroup)`. Please take a look at the `SvgImageBase.GetDrawing` method or the sources of the controls to see how the images are loaded from resources.

**Reply by jaytonic** [reply]
@paulushub I'm not sure that we are talking about the same thing. If you open the SVG I provided, in any browser(tested chrome and edge) or in inkscape, you clearly there is some space:
![image](https://user-images.githubusercontent.com/7582002/119667579-b474d880-be36-11eb-82e5-20eff87e7f1d.png)

Which I don't have when I convert the SVG with your library.

What am I missing?

**Reply by jaytonic** [reply]
in fact, even simplier to demonstrate my issue. With the attached svg(sorry, zipped, github doesn't allow svg upload)
[SingleExpansionCollar.zip](https://github.com/ElinamLLC/SharpVectors/files/6548065/SingleExpansionCollar.zip)


```
<Border BorderBrush="Red">
	<svgc:SvgViewbox Stretch="None"  Source="SingleExpansionCollar.svg" />
</Border>
```

displays this:
![image](https://user-images.githubusercontent.com/7582002/119692552-c90f9b80-be4b-11eb-9000-c6e7ade557da.png)
(please ignore the things that cut the red line, it's other components)

While if I opened this image in Inkscape, I've this:
![image](https://user-images.githubusercontent.com/7582002/119692938-20ae0700-be4c-11eb-9b2e-e31b216b9d0f.png)


Which is what I need, for a lot of different reasons(those images gets aligned with some other connected components on the left-right. And for one components, dependings on its state, I will have differents layers of SVG to display on top of each other).

I tried the other approach you proposed: 

```
<Border BorderBrush="Red" BorderThickness="1">
	<Image x:Name="CallarBottom" Stretch="None" Source="{svgc:SvgImage Source=SingleExpansionCollar.svg}"/>
</Border>

```

but it gives me exactly the same results

**Comment by paulushub** [comment]
@jaytonic OK, setting **EnsureViewboxSize** property to true and turning off the others, could work for your case, but I do not think combining multiple controls will produce the results you want.
Attached is a sample that directly converts the images and add them to a canvas. You can also combine both drawings into a single image and add it to the canvas.
You can try the above options to see the effect on the final drawing and layout.

![shots2](https://user-images.githubusercontent.com/809351/119922299-857f7380-bfaa-11eb-88a6-f4f9f84d2107.png)

[WpfApp4.zip](https://github.com/ElinamLLC/SharpVectors/files/6557809/WpfApp4.zip)

**Reply by jaytonic** [reply]
@paulushub Hi! Thank you very much for the answer.

I edited your example to have the final images(the first one I gave were wrong): And I've the expected behavior:

![image](https://user-images.githubusercontent.com/7582002/119929514-f8bec080-bf7d-11eb-9508-819eb8d777c5.png)

When I enable "EnsureViewBoxSize"(with or without combining diagrams).

Any idea how I can do this in XAML in the SvgViewBox? We have several bindings on each element(top, bottom and main) that will bind the Pen or Brush depending on our business status.
[WpfApp4.zip](https://github.com/ElinamLLC/SharpVectors/files/6558178/WpfApp4.zip)

**Reply by paulushub** [reply]
@jaytonic Sorry for the initial misunderstanding.

I will soon add support for binding Pen and Brush, but I do not think any of the current controls will meet your specific needs.
SharpVectors is designed to be modular and extensible, so you do not have to use any of the supplied controls, just copy any SvgCanvas or SvgViewbox and create a new control that will provide the specific function required by your application.
You can create a new control that has such (TopSource, BottomSource, MainSource etc) as properties 

If doing a new control you meet some limitations of the library, post it as an issue. Any general feature, we add or resolve it at the library level.

**Reply by jaytonic** [reply]
@paulushub I think there is a misunderstanding: We have already using the Pen and brush with bindings, this works nicely in the UserControl we made, that extends SvgViewBox:

```
 public class OverlaySvgViewBox : SvgViewbox
{

	private static readonly Brush _defaultBrush = new SolidColorBrush {Opacity = 0.33, Color = Color.Blue};

	private static readonly Pen _defaultPen = new Pen
	{
		Thickness = 3.0,
		Brush = new SolidColorBrush {Opacity = 0.5, Color = Color.Blue}
	};

	public static readonly DependencyProperty BrushProperty = DependencyProperty.Register(
		"Brush", typeof(Brush), typeof(OverlaySvgViewBox), new PropertyMetadata(_defaultBrush));

	public Brush Brush
	{
		get => (Brush)GetValue(BrushProperty);
		set => SetValue(BrushProperty, value);
	}
	public static readonly DependencyProperty PenProperty = DependencyProperty.Register(
		"Pen", typeof(Pen), typeof(OverlaySvgViewBox), new PropertyMetadata(_defaultPen));

	public Pen Pen
	{
		get => (Pen)GetValue(PenProperty);
		set => SetValue(PenProperty, value);
	}
	protected override void OnSettingsChanged()
	{
		base.OnSettingsChanged();

		IList<Drawing> drawings = ((SvgDrawingCanvas)Child).DrawObjects;
		foreach (GeometryDrawing geometryDrawing in drawings.OfType<GeometryDrawing>())
		{
			Binding brushBinding = new Binding(nameof(GeometryDrawing.Brush))
			{
				Source = this,
				Mode = BindingMode.OneWay
			};
			BindingOperations.SetBinding(geometryDrawing, GeometryDrawing.BrushProperty, brushBinding);

			Binding penBinding = new Binding(nameof(GeometryDrawing.Pen))
			{
				Source = this,
				Mode = BindingMode.OneWay
			};
			BindingOperations.SetBinding(geometryDrawing, GeometryDrawing.PenProperty, penBinding);
		}
	}
}
```

**Then only thing** I'm missing is being able to set this EnsureViewBoxSize when using the SvgViewBox

**Reply by paulushub** [reply]
@jaytonic It is also a current issue: #189. I will expose the setting properties soon.

**Reply by jaytonic** [reply]
Okay, we are not blocked by the pen/brush, but on the EnsureViewBoxSIze, do you already have an issue with this?

**Reply by paulushub** [reply]
I have just checked the sources to the repository, will have to fix other issues before Nuget package.
